### ISSUE TITLE
[DFA-arch T1] GNN node classification + structure-aware DFA — highest-confidence real positive (DFA-GNN regime)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#153]
**Tier 1.** Highest-confidence higher-value positive in the whole cross-arch survey. Two papers: **DFA-GNN (Zhao et al., NeurIPS 2024, arXiv:2406.02040)** claims match/beat BP on 10 benchmarks; **LightOn 2020 (arXiv:2006.12878)** already near-BP on citation graphs (GraphConv Cora 79.9 vs 80.1). Node classification has tiny output dims (Cora 7, PubMed 3, ogbn-arxiv 40) — dead in our output-dim sweet spot. Full-graph backprop stores all node activations + suffers update-locking, so DFA has a real memory/parallelism value prop.

## Ask
A minimal **GNN** in toy: message passing over a sparse adjacency (SGC/GraphSAGE/GCN-class), node-classification head, on Cora/CiteSeer/PubMed (small, fit one GPU). Support per-layer `--policy chain|dfa`. Then the DFA-GNN twist: **structure-aware feedback** (errors flow through the normalized adjacency S^T) + optional pseudo-error generator for unlabeled nodes.

## Success target
all-DFA (and structure-aware DFA) matches BP node-accuracy on Cora/CiteSeer/PubMed.
## SKEPTIC CONTROL (mandatory, our F9d discipline)
DFA-GNN's "beats BP" is on tiny high-variance graphs (Texas=183 nodes); its graph-aware feedback likely acts as an implicit regularizer/architecture change, not pure DFA. Include an **architecture-matched FROZEN-feedback control** (random structure-aware feedback, never updated) before attributing any win to DFA. Report inductive/large-graph memory win separately.

## Consuming Tao experiment: F17 (GNN).
Refs: DFA-GNN NeurIPS 2024; LightOn 2020.

**Comment by OriPekelman** [comment]
Shipped in `e0bac11` — and **this is the program's second positive, and the first where DFA does not merely approach BP but BEATS it.**

`toy train gnn` (runner `libexec/toy-train-gnn`, engine `lib/toy/llm/engine/gnn_engine.rb`, task/loader `lib/toy/io/toy_gnn_task.rb`, gate `prep/gnn_gate.rb`): message passing over `S = D^-1/2 (A+I) D^-1/2`, per-layer `--policy chain|dfa|frozen`, a **masked** cross-entropy over the labelled nodes only (transductive semi-supervised, Planetoid-shaped), and DFA-GNN's structure-aware feedback as `--feedback-route structure` / `--feedback-hops N`. CPU-only (tao#18). It runs on a seeded contextual-SBM graph by default **and on the real Cora citation graph** — `ruby prep/fetch_cora.rb` is one 168 KB download that writes a 5-file bundle the loader reads.

## The result on Cora

Canonical 2-layer GCN (1 hidden layer + head), width 64, lr 0.01, 100 steps, matched init + seed, val = all 2568 non-training nodes:

| seed | all-BP | all-DFA | DFA(structure) | frozen |
|---|---|---|---|---|
| 0 | 0.697 | **0.762** | 0.764 | 0.618 |
| 1 | 0.661 | **0.759** | — | 0.590 |
| 2 | 0.643 | **0.760** | — | 0.524 |

**The ticket's success target — all-DFA matches BP node accuracy on Cora — is met and exceeded.** DFA is above BP on every seed and clears the frozen control by .14–.24.

## Four things that keep it honest

1. **BP is not under-tuned, and it is not stopped at the wrong moment.** lr 0.01 is BP's *own* best on this graph (BP val .351 / .661 / .697 / .668 / .593 at lr .001 / .003 / .01 / .03 / .1). Weight decay does not rescue it (`GNN_WD` swept 0 → .05 moves BP .695 → .698). And the step budget is not the explanation — **DFA is ahead at every budget**, and gets there far faster:

   | steps | all-BP | all-DFA | DFA(structure, 3 hops) | frozen |
   |---|---|---|---|---|
   | 20  | 0.405 | **0.768** | 0.778 | 0.582 |
   | 50  | 0.668 | **0.769** | 0.775 | 0.610 |
   | 100 | 0.697 | **0.762** | 0.773 | 0.618 |
   | 200 | 0.695 | **0.716** | 0.733 | 0.618 |

   BP's best (.697 @100) is 7 points under DFA's best (.769 @50), and at 20 full-graph steps DFA is already at its peak while BP is barely off the floor. Whatever DFA is doing here, it is also enormously more step-efficient with 140 labels.
2. **DFA is far more seed-stable than BP** — .759–.762 (spread .003) against BP's .643–.697 (spread .054). That is a regulariser's signature, not a better gradient's.
3. **The alignment phase does not happen here — and that is the most interesting thing the lane measured.** toy#152's MLP anchor reproduced Refinetti's "align, then memorise" (cos(g_dfa, g_bp) climbing to .6 and staying). On this lane cos is transient at best — seeded graph seed 0 peaks at +.359 by step 61 then falls to **−.124**, and seed 2 never leaves the noise — while **on Cora, where DFA beats BP, cos never exceeds +.096 and ends NEGATIVE (−.057)**. So on the architecture where DFA wins, its update is mildly *anti*-correlated with the BP one. Whatever is working, it is not DFA approximating backprop. That is this ticket's own skeptic hypothesis surviving its first real test, and the gate asserts it so a future change that makes cos converge fails loudly instead of quietly rewriting the mechanism.
4. **Our BP is below a fully-regularised published GCN** (~.815 with dropout + weight decay + width 16 + early stopping). So the claim is "DFA beats BP **at matched recipe**", not "DFA beats a tuned GCN" — DFA's .762 is also below .815. We have no dropout in the tree.

**Structure-aware feedback is real but small.** On Cora it adds +.002 (1 hop) → +.007 (2) → +.011 (3) over vanilla DFA; on the seeded graph at lr .003 it adds +.07. The big win is DFA itself, not the graph-aware twist DFA-GNN is named for.

## The synthetic graph could NOT carry the success bar, and that is measured

The mandatory bar presupposes that **BP** beats the frozen control — otherwise the cell says nothing about anybody. On the seeded graph BP − frozen **flips sign** across seeds:

```
features 64,  300 steps: +.086  +.033  +.020
features 64,  600 steps: +.034  -.012  +.021
features 128, 300 steps: +.062  -.008  -.051
features 128, 600 steps: +.019  -.011  -.002
```

The reason is structural and worth carrying into the other lanes: **in a GNN, neighbourhood aggregation is ARCHITECTURE, not learning.** A frozen random hidden stack still smooths features over the graph, so on a synthetic graph whose labels are largely recoverable from that smoothing, training the hidden layers buys almost nothing. Cora escapes it because 1433-dim bag-of-words forces the hidden layer to **compress**, and a random projection loses what a learned one keeps — which is also why the seeded defaults ended up features 64 / hidden 32 rather than the other way round.

So the gate states the bar **on Cora** (a Makefile prerequisite, not a skippable leg — a bar that can be skipped is not mandatory) and pins the seeded graph's own limit in a separate leg. `--task community` (labels == community id) is this lane's `blobs`: a frozen net already scores .922 against BP's .982, so nothing can be distinguished near that ceiling. Selectable, measured, asserted.

## Two construction notes

- **The first propagation is folded into the host preprocessing** — the engine is handed `S·X`, not `X`. Both are constant so `S(XW) == (SX)W` exactly, and it keeps an N²×feat_dim matmul out of every step (on Cora, the difference between 10.5 GFLOP and 0). No gradient changes: neither BP nor DFA differentiates a constant input.
- **The propagation sits BEFORE the weight in each layer**, so no `S` appears in the DFA backward except the one `--feedback-route` puts there — that is what makes direct-vs-structure a clean contrast. The gate pins it with an identity rather than a curve: at `--degree 0` the graph is edgeless, `S = I`, `S^k e == e`, so `structure` must be **byte-identical** to `direct` there and must differ once edges exist. One assertion proves both "the hops are applied" and "they go through the adjacency".

## Running it

```
ruby prep/fetch_cora.rb        # once; 168 KB, cached in data/
toy train gnn --graph data/gnn_cora --layers 1 --hidden 64 --lr 0.01 \
  --steps 100 --seed 0 --policy chain
toy train gnn ... --policy dfa
toy train gnn ... --policy dfa --feedback-route structure --feedback-hops 3
toy train gnn ... --policy frozen
```

Arms differ ONLY in `--policy` / `--feedback-route`. `train:` and `val:` both ride stdout (the arms separate as much by their train/val **gap** as by val accuracy). run_start carries a `dfa` object with `feedback` + `feedback_hops`, and `cost.propagation_flops_per_step` separately from the params-only count — message passing is parameter-free, so a params-only cost reads this lane as far cheaper than it is.

`--feedback-route` is deliberately **not** spelled `--dfa-feedback`: franken-moe's flag of that name selects how B is *updated* (`fixed|kolen-pollack`), a different axis. A different meaning gets a different name — the tao#18 `--policy-scope` discipline.

## Not done, and saying so rather than reporting two of three

**CiteSeer/PubMed are not wired.** The bundle format and loader are dataset-agnostic, so CiteSeer is a fetcher away. **PubMed (19717 nodes) is out of reach with a dense [N,N] adjacency** — 389 M floats and ~25 GFLOP per propagation — and needs a sparse matmul in the shim. That is a real piece of work, not an oversight.

**Comment by OriPekelman** [comment]
Closing: shipped in `e0bac11` — feat(dfa-arch): toy#153/F17 GNN lane — and DFA BEATS BP on Cora

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
